src/charm.py

```diff
@@ -0,0 +1,469 @@
+
+        self.kafka_cluster = KafkaRequires(
+            self,
+            relation_name=KAFKA_CLUSTER,
+            topic=self.config.topic_name,
+            extra_user_roles=self.config.app_type.value,
+        )
+        self.framework.observe(
+            self.kafka_cluster.on.bootstrap_server_changed, self._on_kafka_bootstrap_server_changed
+        )
+        self.framework.observe(self.kafka_cluster.on.topic_created, self._on_kafka_topic_created)
+        self.framework.observe(self.on[KAFKA_CLUSTER].relation_broken, self._on_relation_broken)
+
+        self.database = DatabaseRequires(
+            self, relation_name=DATABASE, database_name=self.config.topic_name
+        )
+        self.framework.observe(self.database.on.database_created, self._on_database_created)
+        self.framework.observe(
+            self.on[DATABASE].relation_broken, self._on_database_relation_broken
+        )
+
+        self.framework.observe(self.on[CERTIFICATES].relation_joined, self._tls_relation_joined)
+        self.framework.observe(
+            self.certificates.on.certificate_available, self._on_certificate_available
+        )
+
+        # actions
+        self.framework.observe(
+            getattr(self.on, "start_process_action"), self._start_process_action
+        )
+        self.framework.observe(getattr(self.on, "stop_process_action"), self._stop_process_action)
+
+    def _tls_relation_joined(self, event: EventBase) -> None:
+        """Handle the tls relation joined event."""
+        if not self.peer_relation.app_data:
+            event.defer()
+            return
+
+        self.peer_relation.set_private_key(generate_private_key().decode("utf-8"))
+        self._request_certificate()
+
+    def _request_certificate(self):
+        """Request the certificate."""
+        if not self.peer_relation.app_data:
+            return
+
+        private_key = self.peer_relation.app_data.private_key
+        if not private_key:
+            return
+
+        csr = generate_csr(
+            private_key=private_key.encode("utf-8"),
+            subject=os.uname()[1],
+        ).strip()
+
+        self.certificates.request_certificate_creation(certificate_signing_request=csr)
+
+    def _on_certificate_available(self, event: CertificateAvailableEvent) -> None:
+        """Handle the on certificate available event."""
+        if not self.peer_relation.app_data:
+            event.defer()
+            return
+
+        self.write_file(content=self.peer_relation.app_data.private_key, path=KEY_FILE_PATH)
+        self.write_file(content=event.certificate, path=CERT_FILE_PATH)
+        self.write_file(content=event.ca, path=CA_FILE_PATH)
+
+    @staticmethod
+    def write_file(content: str, path: str) -> None:
+        """Write content in a file."""
+        os.makedirs(os.path.dirname(path), exist_ok=True)
+        with open(path, "w") as f:
+            f.write(content)
+
+    def _start_process_action(self, event: ActionEvent) -> None:
```

Comment:
**thought** We might use structured config for validating and casting input parameters. This should be pretty easy and with minimal impact to be done (unlike structured validation for databags :joy: )

Anyhow, this can be done in a separate ticket. I can open a PR about that later when this gets merged as further improvement